File: 51-n-queens/n-queens.java
class Solution {
    public List<List<String>> solveNQueens(int n) {
        List<List<String>> result = new ArrayList<>();
        char[][] board = new char[n][n];
        for (int i = 0; i < n; i++) {
            Arrays.fill(board[i], '.');
        }
        boolean[] col = new boolean[n];
        boolean[] diag1 = new boolean[2 * n - 1]; 
        boolean[] diag2 = new boolean[2 * n - 1]; 
        backtrack(0, n, board, col, diag1, diag2, result);
        return result;
    }

    private void backtrack(int row, int n, char[][] board, boolean[] col, boolean[] diag1, boolean[] diag2,List<List<String>> result) {
        if (row == n) {
            List<String> curr = new ArrayList<>();
            for (int i = 0; i < n; i++) {
                curr.add(new String(board[i]));
            }
            result.add(curr);
            return;
        }
        for (int c = 0; c < n; c++) {
            int d1 = row - c + n - 1;
            int d2 = row + c;

            if (!col[c] && !diag1[d1] && !diag2[d2]) {
                col[c] = diag1[d1] = diag2[d2] = true;
                board[row][c] = 'Q';
                backtrack(row + 1, n, board, col, diag1, diag2, result);
                board[row][c] = '.';
                col[c] = diag1[d1] = diag2[d2] = false;
            }
        }
    }
}

File: 51-n-queens/n-queens.py
class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        res = []
        board = [['.' for _ in range(n)] for _ in range(n)] #empty space
        cols=set()
        next_dia=set() 
        prev_dia=set()

        def backtrack(r):
            if r == n: #if all rows are filled, found valid solution
                copy = ["".join(row) for row in board]
                res.append(copy)
                return
            for c in range(n):
                if c in cols or (r + c) in next_dia or (r - c) in prev_dia:
                    continue
                
                cols.add(c)
                next_dia.add(r + c)
                prev_dia.add(r - c)
                board[r][c] = 'Q'
                
                backtrack(r + 1)
                
                cols.remove(c)
                next_dia.remove(r + c)
                prev_dia.remove(r - c)
                board[r][c] = '.'
                
        backtrack(0)
        return res

